### PR TITLE
Fix GRPO to conform with TRL: Fix loss, make tests accurate, correct metrics computation

### PR DESCRIPTION
**Pull request by shivam15s**
## Summary

Tries to address #626 and other correctness improvements

## Testing Done

- Hardware Type: <BLANK>
- [ ] run `make test` to ensure correctness
- [ ] run `make checkstyle` to ensure code style
- [ ] run `make test-convergence` to ensure convergence

### CODE REVIEW

**Comment by qingquansong** [comment]
One more thing I forget to mention in the issue. The aggregation way is also implemented differently as the original GRPO paper equation. It should be a per sequence/response avg and then per group avg. Here we directly aggregate the entire batch loss and average based on sum of tokens. Ideally it should be a three level thing (or maybe two level depending on the input) `batch avg <- group avg <- seq/response avg`

File: src/liger_kernel/chunked_loss/grpo_loss.py [hunk 1]
```diff
@@ -7,60 +7,60 @@ class LigerFusedLinearGRPOFunction(LigerFusedLinearRLHFBase):
...
-
-        # Combine losses
-        per_token_loss = policy_loss + beta * kl_div
-
-        # Apply masking and normalize
-        masked_loss = per_token_loss * attention_mask
-        seq_lengths = attention_mask.sum()
-        seq_lengths = torch.clamp(seq_lengths, min=1.0)
-        loss = masked_loss.sum() / seq_lengths
+        old_per_token_logps = old_per_token_logps if old_per_token_logps is not None else per_token_logps.detach()
```
Comment (qingquansong) [inline src/liger_kernel/chunked_loss/grpo_loss.py:38]: It should be great if we can create two small utility functions for the kl compute part and also the advantage clip part and can be reused and switched by different losses. For example, GRPO use k3 approximation here for the kl penalty (and you can use k3 to represent the function name) and we can switch to others if needed. Thank you1

**Review by qingquansong — DISMISSED** [review]
LGTM! Just two minor small comments!

File: src/liger_kernel/chunked_loss/fused_linear_rlhf.py [hunk 2]
```diff
@@ -158,60 +229,74 @@ def accumulate_chunk(input_chunk, attention_mask_chunk, rewards_chunk, ref_input
...
+            logits = logits / temperature
 
         # Compute log probabilities using softmax over the last dimension
         log_probs = F.log_softmax(logits.float(), dim=-1)
 
-        # Monitoring: compute mean of logits
-        batch_size, seq_len, _ = input_chunk.shape
-        logits_mean = logits.sum() / (batch_size * seq_len * weight.shape[0])
-        return log_probs, logits, logits_mean
+        return log_probs, logits
```
Comment (lancerts) [inline src/liger_kernel/chunked_loss/fused_linear_rlhf.py:299]: So this is a backward-breaking change? What was the motivation?

Comment (shivam15s) [reply src/liger_kernel/chunked_loss/fused_linear_rlhf.py:299]: logits_mean isn't needed for any metric logging as far as trl is concerned

File: src/liger_kernel/chunked_loss/grpo_loss.py [hunk 3]
```diff
@@ -7,60 +7,60 @@ class LigerFusedLinearGRPOFunction(LigerFusedLinearRLHFBase):
...
         # Calculate metrics
-        metrics = (
-            chosen_token_logprobs.mean(),  # mean log prob
-            chosen_token_logprobs.std(),  # std log prob
-            log_probs.mean(),  # mean all log probs
-            ((kl_div * attention_mask).sum(dim=1) / attention_mask.sum(dim=1)).mean(),  # mean KL div
+        metrics = []
+        if beta != 0.0:
+            metrics.append(((kl_div * attention_mask).sum() / torch.clamp(full_attention_mask.sum(), min=1.0)))
+        is_clipped = ((coef_1 < 1 - epsilon_low) & (advantages.unsqueeze(1) < 0)) | (
```
[the remark this answers is not in the document]
Comment (shivam15s) [reply src/liger_kernel/chunked_loss/grpo_loss.py:60]: I feel this is because of how we compute per_token_loss:
`per_token_loss = -torch.min(per_token_loss1, per_token_loss2)`
Assume, `coef_1 = 1-2*eps; and coef_2 = 1 - eps;` 
if advantage is `+ve`, `torch.min()` would select `per_token_loss_1` rather than `per_token_loss2` as `per_token_loss1` < `per_token_loss2`.
So, even though `coef_2` is clipped, we still took `per_token_loss_1` as loss.

tldr: sign of advantage is important to figure out is_clipped

**Review by qingquansong — APPROVED** [review]
LGTM! :shipit:

**Review by qingquansong — APPROVED** [review]
